src/main/java/org/eclipse/uprotocol/cloudevent/factory/UCloudEvent.java

```diff
@@ -351,5 +352,76 @@ static UMessageType getMessageType(String ce_type){
                 return UMessageType.UMESSAGE_TYPE_UNSPECIFIED;
         }
     }
+    /**
+     * Get the UMessage from the cloud event
+     * @param event The CloudEvent containing the data.
+     * @return returns the UMessage
+     */
+    static UMessage toMessage(CloudEvent event) {
+        UUri source = LongUriSerializer.instance().deserialize(getSource(event));
+        UPayload payload = UPayload.newBuilder().setFormat(UPayloadFormat.UPAYLOAD_FORMAT_PROTOBUF)
+                .setValue(getPayload(event).toByteString()).build();
+        UAttributes.Builder builder =
+                UAttributes.newBuilder().setId(LongUuidSerializer.instance().deserialize(event.getId()))
+                        .setType(getMessageType(event.getType())).setTtl(getCommunicationStatus(event))
+                        .setPriority(UPriority.valueOf(getPriority(event).toString()));
+        if (getSink(event).isPresent()) {
+            builder.setSink(LongUriSerializer.instance().deserialize(getSink(event).get()));
+        }
+        if (getTtl(event).isPresent()) {
+            builder.setTtl(getTtl(event).get());
+        }
+        if (getRequestId(event).isPresent()) {
+            builder.setReqid(LongUuidSerializer.instance().deserialize(getRequestId(event).get()));
+        }
+        if (getToken(event).isPresent()) {
+            builder.setToken(getToken(event).get());
+        }
+        Integer permission_level = extractIntegerValueFromExtension("permission_level", event).orElse(UCode.OK_VALUE);
```

Comment:
you'll need a check around this as permission_level is optional

Reply:
Done